src/services/user.service.ts

```diff
@@ -0,0 +1,82 @@
+import { Model } from "sequelize";
+
+import UserModel from "../models/user.model";
+import User from "../types/user.type";
+
+import { userAccess } from "../data-access";
+
+interface UserService {
+  userModel: typeof UserModel;
+}
+
+const { userLogin, findUser, autoSuggestions } = userAccess;
+
+class UserService {
+  constructor(userModel: typeof UserModel) {
+    this.userModel = userModel;
+  }
+
+  async login(login: User["login"], password: User["password"]) {
+    const user: Model<User> | null = await this.userModel.findOne(
+      userLogin(login, password)
+    );
+
+    if (!user) return null;
+
+    return user;
+  }
+
+  async createUser(userParams: Partial<User>) {
+    userParams.isDeleted = false;
+
+    const user = await this.userModel.create(userParams);
+    return user;
+  }
+
+  async updateUser(userId: User["id"], userParams: Partial<User>) {
+    const user = await this.userModel.findOne(findUser(userId));
+
+    if (!user) return null;
+
+    if (userParams.login) {
+      await user.update({ login: userParams.login });
```

Comment:
It's always better to do a one big update than multiple consecutive small one.
Otherwise you might exhaust the connections at some point and the performance of doing a single round trip to the DB and back is much better.
Some DBs even have a concept of a Bulk Operation when several unrelated DB operations are executed in a bulk to achieve the best performance results, kind of like Promise.all for DBs.

Reply:
Good point - I didn't consider this. Previously I'd mostly just do incremental updates - I didn't consider this might create a bottle neck.